libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/LedgerTypes/Tests.hs

```diff
@@ -0,0 +1,362 @@
+-- specified. We use the strategy of passing (Term fn x) as inputs to those specifcations.
+-- For example, the AccountState must have sufficient capacity to support the InstantaneousRewards
+-- So we pass a (Term fn AccountState) as input to 'instantaneousRewardsSpec'
+-- In order to create tests, we need specifications that are fully applied.
+-- The idea is to combine several Specifications to get a Gen(composed specifations)
+-- For example (dstateSpec @Shelley !$! accountStateSpec !*! poolMapSpec)
+-- is a (Gen (Specification ConwayFn (DState Shelley)))
+-- If a Specification takes an actual PParams (not a (Term fn PParams)), like
+-- lederstateSpec, we can combine it like this using the Functor <$>, rather than our !$!
+-- (ledgerStateSpec <$> genConwayFn pparamsSpec !*! accountStateSpec !*! epochNoSpec)
+-- ====================================================================================
+
+-- GenFromSpec fixed at ConwayFn
+genConwayFn :: (HasCallStack, HasSpec ConwayFn a) => Specification ConwayFn a -> Gen a
+genConwayFn = genFromSpec @ConwayFn
+
+-- Analagous to <$> except the function to be applied takes a (Term fn a -> t) instead of (a -> t)
+infixr 6 !$!
+(!$!) ::
+  forall fn t a.
+  (BaseUniverse fn, HasSpec fn a) =>
+  (Term fn a -> t) -> Specification fn a -> Gen t
+(!$!) bf specA = do a <- genFromSpec @fn @a specA; pure (bf (lit a))
+
+-- Analagous to <*> except the function to be applied takes a Gen (Term fn a -> t) instead of F (a -> t)
+infixl 4 !*!
+(!*!) ::
+  forall fn t a.
+  (BaseUniverse fn, HasSpec fn a) =>
+  Gen (Term fn a -> t) -> Specification fn a -> Gen t
+(!*!) gentf specA = do a <- genFromSpec @fn @a specA; f <- gentf; pure (f (lit a))
+
+poolMapSpec ::
+  Specification ConwayFn (Map (KeyHash 'StakePool StandardCrypto) (PoolParams StandardCrypto))
+poolMapSpec = hasSize (rangeSize 8 8)
+
+delegationsSpec ::
+  Specification
+    ConwayFn
+    (Map (Credential 'Staking StandardCrypto) (KeyHash 'StakePool StandardCrypto))
+delegationsSpec = (hasSize (rangeSize 8 12))
+
+-- ====================================================================
+-- HSpec tests
+-- ===================================================================
+
+soundSpec ::
+  forall t. (HasSpec ConwayFn t, PrettyA t) => Gen (Specification ConwayFn t) -> Gen Property
+soundSpec specGen = do
+  spect <- specGen
+  x <- genConwayFn @t spect
+  pure $
+    property $
+      counterexample (show ("Does not meet spec\n" <> prettyA x)) (conformsToSpec x spect)
+
+soundSpecWith ::
+  forall t.
+  (HasSpec ConwayFn t, PrettyA t) =>
+  Int -> Gen (Specification ConwayFn t) -> SpecWith (Arg Property)
+soundSpecWith n specx = it (show (typeRep (Proxy @t))) $ withMaxSuccess n $ property $ (soundSpec @t specx)
+
+utxoStateGen :: forall era. LedgerEra era ConwayFn => Gen (Specification ConwayFn (UTxOState era))
+utxoStateGen =
+  utxoStateSpec @era
+    <$> genConwayFn @(PParams era) pparamsSpec
+    <*> (lit <$> wff @(CertState era) @era)
+
+spec :: Spec
```

Comment:
I think this test suite could be made era-parametric and then called for each era. That way we don't have to copy-paste each type for each era

Reply:
Yes, that is a good idea.  I will give it a try.